_posts/2025-04-19-navigating-pagination-with-hybrid-query.md

```diff
@@ -0,0 +1,229 @@
+
+OpenSearch continues to enhance its hybrid query capabilities with features like post-filters, aggregations, query parallelization, explain, and sorting. Building on this momentum, OpenSearch 2.19 introduces pagination support through the new 'pagination_depth' parameter. When used with traditional from and size parameters, this enhancement helps you efficiently navigate through hybrid query result sets. For example, you can now paginate through thousands of search results one page at a time.
+
+In this blog post, we'll explain pagination and the `pagination_depth` parameter. We'll explore why pagination depth is necessary for hybrid queries, show you how to use this feature, and share benchmarking results.
+
+## Understanding pagination
+
+OpenSearch uses pagination to divide large result sets into manageable pages. Each page contains the number of search results specified by the size parameter in your search request. OpenSearch supports four pagination techniques:
+- `from` and `size` parameters
+- Scroll search operation
+- `search_after` parameter
+- Point in Time with `search_after`
+
+For more information about these techniques, see the [OpenSearch documentation](https://opensearch.org/docs/latest/search-plugins/searching-data/paginate/).
+
+This blog focuses on using the `from` and `size` parameters with the new `pagination_depth` parameter in hybrid queries.
+
+## How pagination_depth works
+
+The `pagination_depth` parameter specifies the maximum number of search results to retrieve from each shard for every subquery. This gives you precise control over how many results to hybridize for each subquery.
+
+The following figure presents an index with three shards. The example hybrid query search request includes two subqueries (a match query and a k-NN query), with `pagination_depth = 20`, `size = 10`, and `from = 5`.
+
+![pagination-depth](/assets/media/blog-images/2025-04-19-navigating-pagination-with-hybrid-query/pagination-depth.png){:class="img-centered"}
+
+Here's how this example works:
+
+1. With `pagination_depth` set to 20, each shard retrieves up to 20 results for both the match query and the k-NN query (this example assumes that each shard contains at least 20 results per query.)
+2. The coordinator node receives 120 total results (60 for the match query and 60 for the k-NN query) and processes them using hybrid search techniques, including normalization and score combination.
+3. When applying hybrid search techniques, any duplicate results appearing in both subqueries are merged into a single result.
+4. The final results are trimmed according to top 10 results ( because of `size = 10`), starting from the 5th entry (because of `from = 5`).
```

Comment:
```suggestion
4. The final results are trimmed according to the top 10 results (because of `size = 10`), starting from the fifth entry (because of `from = 5`).
```